File: src/main/java/net/sf/rails/util/SequenceUtil.java
package net.sf.rails.util;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

import net.sf.rails.game.Company;
import net.sf.rails.game.Player;
import net.sf.rails.game.PublicCompany;
import net.sf.rails.game.financial.Bank;
import net.sf.rails.game.state.MoneyOwner;



public class SequenceUtil {

    // private constructor
    private SequenceUtil() {};
    
    private static <E extends MoneyOwner> List<E> 
        selectMoneyOwners(Class<E> clazz, Collection<MoneyOwner> coll)  {
        
        // select all cashholders of that type
        List<E> list = new ArrayList<E>();
        
        for (MoneyOwner c:coll) { 
            if (clazz.isAssignableFrom(c.getClass())) {
                @SuppressWarnings("unchecked")
                E cast = (E) c;
                list.add(cast);
            }
        }
        
        return list;
    }
    
    
    /**
     * Defines a sorting on cashHolders
     * @return sorted list of cashholders
     */
    public static List<MoneyOwner> sortCashHolders(Collection<MoneyOwner> coll) {
        
       List<MoneyOwner> sortedList = new ArrayList<MoneyOwner>();

       // first add players
       List<Player> players = selectMoneyOwners(Player.class, coll);
       Collections.sort(players);
       sortedList.addAll(players);
       
       // then public companies
       List<PublicCompany> PublicCompanys = selectMoneyOwners(PublicCompany.class, coll);
       Collections.sort(PublicCompanys, Company.COMPANY_COMPARATOR);
       sortedList.addAll(PublicCompanys);
       
       // last add the bank
       sortedList.addAll(selectMoneyOwners(Bank.class, coll));
       
       return sortedList;
    }
}

File: dotnet/RailsLib.Net/Net/Util/SequenceUtil.cs
using GameLib.Net.Game;
using GameLib.Net.Game.Financial;
using GameLib.Net.Game.State;
using System;
using System.Collections.Generic;
using System.Text;

namespace GameLib.Net.Util
{
    public static class SequenceUtil
    {
        private static List<T> SelectMoneyOwners<T>(IEnumerable<IMoneyOwner> coll) where T : IMoneyOwner
        {

            // select all cashholders of that type
            List<T> list = new List<T>();

            foreach (IMoneyOwner c in coll)
            {
                if (typeof(T).IsAssignableFrom(c.GetType()))
                {
                    T cast = (T)c;
                    list.Add(cast);
                }
            }

            return list;
        }

        /**
         * Defines a sorting on cashHolders
         * @return sorted list of cashholders
         */
        public static List<IMoneyOwner> SortCashHolders(ICollection<IMoneyOwner> coll)
        {
            List<IMoneyOwner> sortedList = new List<IMoneyOwner>();

            // first add players
            List<Player> players = SelectMoneyOwners<Player>(coll);
            players.Sort();
            sortedList.AddRange(players);

            // then public companies
            List<PublicCompany> PublicCompanys = SelectMoneyOwners<PublicCompany>(coll);
            PublicCompanys.Sort(CompanyComparer.COMPANY_COMPARATOR);
            sortedList.AddRange(PublicCompanys);

            // last add the bank
            sortedList.AddRange(SelectMoneyOwners<Bank>(coll));

            return sortedList;
        }
    }
}
